qcodes/data/data_set.py

```diff
@@ -695,9 +703,31 @@ def get_arrays(
             action_indices: All array action indices must start with the provided
                 indices.
             set_arrays: Whether to include setpoint arrays
+            data_group: Optional data group that the array should belong to
         """
         arrays = sorted(self.arrays.values(), key=lambda arr:arr.action_indices)
 
+        # Filter by data group
+        if data_group is not None:
+            if 'data_groups' not in self.metadata:
+                warn('Metadata does not contain data groups')
+            else:
+                valid_idxs = [
+                    tuple(key) for key, val in self.metadata['data_groups']
+                    if data_group in val
```

Comment:
I think this should be an explicit equality, otherwise you might get key clashes, e.g. a pair of data groups called 'ESR_Rabi' and 'ESR_spectrum'.
```suggestion
                    if data_group == val
```